### PR TITLE
feat: impl visitor pattern for rendering

### LINKED ISSUE #87: [REQ] Use visitor pattern for the `Render` trait
At the moment, every element implements the defined `Render` trait.
This makes it impossible to define custom render types not defined in the `Render` trait.

Using the [visitor-pattern](https://rust-unofficial.github.io/patterns/patterns/behavioural/visitor.html), the `Render` trait would contain a function for each Unimarkup base type.
A new render type can then be supported by implementing each function of the `Render` trait.

This new approach would easily lead to breaking changes when adding new base types.
To prevent this, every function must have a default implementation.

e.g. how the trait might look like:

```rust
trait Render<'a, T> {
  type State;

  fn render_document(&mut self, &'a item: Document, &'a mut context: Context<State>) -> T;
  fn render_paragraph(&mut self, &'a item: Paragraph, &'a mut context: Context<State>) -> T;
  // ...
}
```

In the above example, a context is given in addition to the item, to provide access to defined macros, memorables, references, ...
The `State` type offers the possibility for the custom renderer to keep a custom state across rendering.
For PDF, this could be the PDF document itself.

### PR DESCRIPTION
**Pull request by mhatzl**
This PR closes #87 

Instead of specifying different output formats in the `Render` trait, the new approach defines a `Renderer` trait to set render functions for each Unimarkup type. This allows implementations to write custom output formats, while reusing the Unimarkup reference parser.

To clarify crate naming, I moved the code from `unimarkup-core` to `unimarkup-parser`, since the current code is mainly about parsing and generating an AST (Document).

With the move to the new rendering approach, the `unimarkup-render` crate has a dependency on `parser`.
Therefore, `parser`, and any crate `parser` depends on, cannot set `render` as dependency.
To still provide a single entry-lib that combines parsing and rendering, the new `core` crate provides a `Unimarkup` struct that may be used to `parse` and `render` Unimarkup content to any format implementing the new `Renderer` trait.
In addition, I re-exported all `unimarkup-*` crates used in `core` so users only have to add `core` as dependency.
At the moment,`inline` is not re-exported, since it is not set as dependency, but we might change that at a later point.

### CODE REVIEW

**Comment by mhatzl** [comment]
I also moved the `output` config from the preamble to the general config struct as internally discussed.
Having formats set in the preamble makes it harder for us to handle custom formats, and also makes preamble merging much harder and confusing.

File: cli/src/main.rs [hunk 1]
```diff
@@ -4,8 +4,8 @@ use unimarkup_commons::config::{Config, ConfigFns};
 
 use crate::log_id::{GeneralError, GeneralInfo};
 
+mod compile;
```
Comment (nfejzic) [inline cli/src/main.rs:7]: Maybe call this module `compiler`?

File: core/src/lib.rs [hunk 2]
```diff
@@ -1,10 +1,60 @@
...
+    pub fn get_formats(&self) -> impl Iterator<Item = &OutputFormatKind> {
+        self.doc.output_formats()
+    }
+
+    pub fn render<T: OutputFormat>(&self, renderer: impl Renderer<T>) -> Result<T, RenderError> {
+        unimarkup_render::render::render(&self.doc, renderer)
+    }
+
+    pub fn render_html(&self) -> Result<Html, RenderError> {
+        unimarkup_render::render::render(&self.doc, HtmlRenderer::default())
```
Comment (nfejzic) [inline core/src/lib.rs:53]: We could also call the `render` function defined on `Unimarkup` here, but both options are ok for me. 

Just to see how it looks like: 

```suggestion
        self.render(HtmlRenderer::default())
```

File: core/src/lib.rs [hunk 3]
```diff
@@ -1,10 +1,60 @@
...
+    }
+}
+
+pub struct RenderedOutput {
+    pub content: String,
+    pub kind: OutputFormatKind,
+}
```
Comment (nfejzic) [inline core/src/lib.rs:60]: I didn't find any reference to this, can we delete it? 

```suggestion
```

File: core/src/lib.rs [hunk 4]
```diff
@@ -1,10 +1,60 @@
...
+        if um_content.is_empty() {
+            return Unimarkup {
+                doc: Document {
+                    blocks: vec![],
+                    config,
+                    ..Default::default()
+                },
+            };
+        }
+
```
Comment (nfejzic) [inline core/src/lib.rs:34]: Is this necessary? Wouldn't the parser finish early returning document with empty vec when content is empty? If so maybe we could spare ourselves this part.

**Review by nfejzic — COMMENTED** [review]
I'm very happy with this. I like how rendering is now completely generic 💪🏻👍🏻. Left some comments for potential improvement and discussion.

**Comment by nfejzic** [comment]
One more thing, some functions in `Renderer` trait enable the implementer to simply implement everything in that function (like `render_blocks, render_block, render_inlines, render_inline`). 

We could prevent this by splitting `Renderer` into two traits. One trait would be internal and can only be implemented by us. The other trait is meant for custom implementations, something like this: 

```rust
mod private {
    pub trait Sealed {}
    impl<T> Sealed for T where T: Renderer {} 
}

pub trait RendererSealed<T: OutputFormat>: Sealed {
    fn render_blocks(&mut self, blocks: &[Block], context: &Context) -> Result<T, RenderError>;
    fn render_block(&mut self, block: &Block, context: &Context) -> Result<T, RenderError>;
    fn render_inlines(&mut self, inlines: &[Inline], context: &Context) -> Result<T, RenderError>;
    fn render_inline(&mut self, inline: &Inline, context: &Context) -> Result<T, RenderError>;
}

impl<T, O: OutputFormat> RendererSealed<O> for T
where
    T: Renderer<O>
{
    fn render_blocks(&mut self, blocks: &[Block], context: &Context) -> Result<T, RenderError> { 
        // ... 
    }
    fn render_block(&mut self, block: &Block, context: &Context) -> Result<T, RenderError> { 
        // ... 
    }
    fn render_inlines(&mut self, inlines: &[Inline], context: &Context) -> Result<T, RenderError> { 
        // ... 
    }
    fn render_inline(&mut self, inline: &Inline, context: &Context) -> Result<T, RenderError> { 
        // ... 
    }
}

// the Renderer trait is now left only with specific functions like render_paragraph etc.
```

This comes with a trade-off! Some output cannot correctly render without having an insight into the output rendered until that specific point. With custom implementation of `render_blocks` the implementer has full control over output and can correctly implement that. 

There are ways to circumvent the problem. For example, implementer can have the rendered content as internal state of custom renderer. Render functions modify this internal state, and return the up-to that point rendered document (for example by cloning, or possibly by using some smart pointer with interior mutability). The `append` function of the output format can then be implemented in a way that it takes advantage of that information. For example, if smart pointer is used with interior mutability, it would be possible to produce output once, and then append function simply does nothing, because the content is already updated (smart pointer points to same location as the one in the renderer state).

File: render/src/html/render.rs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        // };
+        let language = "auto";
+
+        let html = Html::new_with_body(HtmlElement {
+            name: "div".to_string(),
+            attributes: HtmlAttributes::default(),
+            content: Some(highlight::highlight_html_lines(
+                &verbatim.content,
+                language,
+                DEFAULT_THEME,
```
[the remark this answers is not in the document]
Comment (mhatzl) [reply render/src/html/render.rs:61]: Looking into `syntect`, I think we will need to write our own highlighting functionality.
Using `syntect`, we would need to set a theme during rendering (which we currently do), but there is no theme that is able to switch between light and dark styling. Since light/dark preference depends mostly on the client browser, we cannot use these themes. So there is also no reason to put the default theme into the context struct.

tldr: highlighting/theming must change. we need to write our own highlighter on top of parsing from `syntect`

File: render/src/html/render.rs [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+        context: &Context,
+    ) -> Result<Html, crate::log_id::RenderError> {
+        let inner = self.render_nested_inline(italic, context)?;
+
+        Ok(Html::new_nested("em", HtmlAttributes::default(), inner))
+    }
+
+    fn render_underline(
+        &mut self,
+        underline: &unimarkup_inline::NestedContent,
```
[the remark this answers is not in the document]
Comment (mhatzl) [reply render/src/html/render.rs:90]: sounds good, but I think this deserves its own PR.

File: render/src/html/mod.rs [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        html
+    }
+}
+
+impl OutputFormat for Html {
+    fn new(_context: &crate::render::Context) -> Self {
+        Html {
+            head: HtmlElements(Vec::new()),
+            body: HtmlElements(Vec::new()),
+            lang: "en-US".to_string(),
```
[the remark this answers is not in the document]
Comment (mhatzl) [reply render/src/html/mod.rs:72]: that is correct, I just did not know how to setup the context structure.
should the context struct contain a reference to the `Document` struct, or should it mirror some fields of the `Document`?

File: render/src/html/mod.rs [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+#[derive(Debug, Default)]
+pub struct HtmlAttributes(Vec<HtmlAttribute>);
+#[derive(Debug, Default)]
+pub struct HtmlElements(Vec<HtmlElement>);
+
+#[derive(Debug, Default)]
+pub struct Html {
+    pub head: HtmlElements,
+    pub body: HtmlElements,
```
[the remark this answers is not in the document]
Comment (mhatzl) [reply render/src/html/mod.rs:31]: thanks. nice catch 👍

Just not sure about creating custom Head and Body structs.
They won't have attributes, because Unimarkup syntax does not allow it.
And `accidentally` passing something else won't be solved by this either, since both accept `HtmlElements`.

File: parser/src/elements/atomic/heading.rs [hunk 9]
```diff
@@ -190,88 +188,66 @@ impl AsRef<Self> for Heading {
...
+//             let html = heading.render_html();
+
+//             let expected = format!(
+//                 "<h{} id='{}'><pre><code>This</code></pre> <em>is <sub>a</sub></em> <strong>heading</strong></h{}>",
+//                 level, id, level
+//             );
+//             assert_eq!(html.body, expected);
+//         }
+//     }
+// }
```
[the remark this answers is not in the document]
Comment (mhatzl) [reply parser/src/elements/atomic/heading.rs:253]: I agree.
Just kept the tests to remember what content to use in our snapshot tests

File: core/src/lib.rs (hunk 3 above)
Comment (mhatzl) [reply core/src/lib.rs:60]: ah nice catch.
It is a remainder of some general render functions I removed.

File: core/src/lib.rs (hunk 4 above)
Comment (mhatzl) [reply core/src/lib.rs:34]: not sure 😄 
I think in a previous version we had a panic parsing empty content.
If this is not the case anymore, I will remove it here.

**Comment by mhatzl** [comment]
Thanks for the review.
I created a new discussions [entry](https://github.com/unimarkup/unimarkup-rs/discussions/93) about the `Renderer` trait flexibility. For now, I would keep it as is.

File: render/src/html/render.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mhatzl) [reply render/src/html/render.rs:61]: I think we might be able to use [`ClassedHTMLGenerator`](https://docs.rs/syntect/latest/syntect/html/struct.ClassedHTMLGenerator.html) from `syntect`.

They also provide the possibility to convert themes to CSS, so we might include light and dark media queries for one light and one dark theme.

[the remark this answers is not in the document]
Comment (mhatzl) [reply render/src/html/render.rs:61]: should I change it already in for this PR?

[the remark this answers is not in the document]
Comment (nfejzic) [reply render/src/html/render.rs:61]: If possible I would suggest implementing it as part of this PR. Is there option in the config to choose theme? If so, we could probably add the option to the context.

One suggestion regarding context, it would probably make the most sense to getter methods on `Context`, instead of exposing the fields of `Context` directly. This would make it possible to change `Context` in a backwards-compatible manner. 

For example: 

```rust
impl Context {
    pub fn code_block_style(&self) -> T { // or code_block_theme or whatever fits
        // Return whatever the T is
    }

    // ...
```

File: render/src/html/render.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (nfejzic) [reply render/src/html/render.rs:90]: Of course, I already hinted that. I will open an issue.

File: render/src/html/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (nfejzic) [reply render/src/html/mod.rs:72]: See comment above, basically I would make all fields of `Context` private, and expose all of its options/data through methods. This makes it possible to change the structure `Context` without introducing breaking changes (same method still returns same thing, even though underlying data structure changed).

File: render/src/html/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (nfejzic) [reply render/src/html/mod.rs:31]: I meant, accidentally passing some other `HtmlElement` where we meant `HtmlBody` for example. We can remove attributes, but I would suggest still creating separate elements with newtype pattern. 

It's really just a way to encode in type system that we want that specific element. In that case, compiler helps if implementation of `Display` is missing for `HtmlHead` for example.

**Comment by nfejzic** [comment]
> Thanks for the review. I created a new discussions [entry](https://github.com/unimarkup/unimarkup-rs/discussions/93) about the `Renderer` trait flexibility. For now, I would keep it as is.

I agree, I guess this is something where feedback of crate users would be very helpful. Both options are fine, one of them might be *the better one*, I just don't know what *better* would mean in this case. Thanks for opening the discussion.

**Comment by mhatzl** [comment]
I adapted syntax highlighting to use CSS classes, and added a stylesheet that uses light/dark themes for highlighting depending on user preferences.

The set language is now also set in the rendered HTML output.
This must change when we add internationalization support, since then we need to distinguish between set languages and the currently active one.

File: render/src/html/mod.rs [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+
+impl Html {
+    pub fn with_head(head: HtmlHead) -> Self {
+        Html {
+            head,
+            ..Default::default()
+        }
+    }
+
+    pub fn with_body(element: HtmlElement) -> Self {
```
Comment (nfejzic) [inline render/src/html/mod.rs:61]: Should this parameter be `HtmlBody`? Or is there a reason `HtmlElement` is used?

**Review by nfejzic — APPROVED** [review]
Looks nice! Left a single comment for clarification, otherwise we can merge this! 👍🏻

File: render/src/html/mod.rs (hunk 10 above)
Comment (mhatzl) [reply render/src/html/mod.rs:61]: Since `HtmlBody` only contains `HtmlElements`, it felt tedious to use `HtmlBody` as parameter.
But it is probably better to change this to `Vec<HtmlElement>`.

Comment (nfejzic) [reply render/src/html/mod.rs:61]: In that case I would suggest implementing `From<HtmlElement>` or `From<Vec<HtmlElement>`, or renaming this function to `with_elements`.
Function name `with_body` is misleading in this case.

File: render/src/html/render.rs [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+    }
+
+    fn render_superscript(
+        &mut self,
+        superscript: &unimarkup_inline::NestedContent,
+        context: &Context,
+    ) -> Result<Html, crate::log_id::RenderError> {
+        let inner = self.render_nested_inline(superscript, context)?;
+
+        Ok(Html::nested("sup", HtmlAttributes::default(), inner))
```
Comment (nfejzic) [inline render/src/html/render.rs:122]: Another improvement we could have is to use `enum HtmlTag` instead of strings for tags. I'll leave it up to you to decide whether we should do it in this PR or in another one (or at all).

**Review by nfejzic — DISMISSED** [review]

File: render/src/html/render.rs (hunk 11 above)
Comment (mhatzl) [reply render/src/html/render.rs:122]: good idea. I will implement it in this PR

**Review by nfejzic — APPROVED** [review]
